### PR TITLE
machine: Add SSHConnection.reboot()

### PR DESCRIPTION
**Pull request by martinpitt**
Cockpit's tests repeat this code a lot, so let's add a proper API for
it.

---

See `git grep m.spawn.*reboot test/verify` in cockpit.git.

### CODE REVIEW

**Comment by martinpitt** [comment]
Used/tested by https://github.com/cockpit-project/cockpit/pull/17089

**Review by allisonkarlitskaya — COMMENTED** [review]
Why the `sync` and the `sleep`?

I understand that this is just copying existing usage, but if we're making an API, probably it makes sense to think about why this is necessary.  They both look hella suspicious.

**Review by marusak — APPROVED** [review]

**Comment by martinpitt** [comment]
Right, I already reduced the `sync && sync && sync` to a single `sync`, but it really shouldn't be necessary -- rebooting already syncs. So I'll take it out.

The "sleep" is a race condition. It needs to postpone the reboot long enough for the m.execute() to return, so that it doesn't prematurely kill SSH.

**Comment by allisonkarlitskaya** [comment]
> The "sleep" is a race condition. It needs to postpone the reboot long enough for the m.execute() to return, so that it doesn't prematurely kill SSH.

Can we find another way to do that?

**Comment by martinpitt** [comment]
If you have a better idea, I'm all ears.

**Comment by allisonkarlitskaya** [comment]
`check=False` parameter to `machine.execute()`?

**Comment by martinpitt** [comment]
I don't like that. That would just hide errors, in case we break something in the VM so that rebooting would fail.

**Comment by martinpitt** [comment]
I mean, we *can* try to drop the sleep entirely, but it would still be fundamentally the same race condition. But this never got in the way.

**Comment by allisonkarlitskaya** [comment]
I actually don't understand why the sleep makes any difference at all here.  It just delays the `reboot` call.  It doesn't cause the ssh connection to terminate any quicker relative to the `reboot` call.  It still waits until after it's invoked.

**Comment by allisonkarlitskaya** [comment]
Also: why do we pass both a `command` and `script` parameter?  It seems `script` gets ignored in the case that `command` is given...

**Comment by allisonkarlitskaya** [comment]
> @allisonkarlitskaya : It does delay it -- this is m.spawn() (asynchronous), _not_ m.execute()

Right.  Thanks for the clarification.

**Comment by allisonkarlitskaya** [comment]
What about something like `cat && reboot`?  ie: block until stdin is closed and then proceed.  It might even make sense to bake that into the shell stuff in the implementation of `spawn()` in that case...

**Comment by martinpitt** [comment]
OK, moving to draft for later. This really isn't the bit that I wanted to do right now, and it's not at all important.

**Comment by allisonkarlitskaya** [comment]
So using this in the shell:

```
ps -o ppid= $$
```

gets you the ppid, which is equal to the sshd instance that was forked off to handle the current connection.  We just need to wait for that process to be gone.




> I don't like that. That would just hide errors, in case we break something in the VM so that rebooting would fail.

The logic here is false: the entire point of this is that we don't call `reboot` until after the ssh session closes, which means that we won't catch errors.

**Comment by martinpitt** [comment]
I experimented with `cat`, and that seems to work well enough -- we don't actually close stdin in execute anywhere, unless we give a stdin input string. Doing a run against that in https://github.com/cockpit-project/cockpit/pull/17089

**Comment by allisonkarlitskaya** [comment]
I'm not super keen on the `cat` idea because it depends on the implementation of `sshd` and could just be trading one race for another.

I really start to think the correct solution here is to give a way to ignore `execute()`/`spawn()` failures...

**Comment by allisonkarlitskaya** [comment]
But: I feel like we've gone around enough times and thought about it enough that I'm happy to stand out of the way and let you decide which approach you prefer.

**Comment by martinpitt** [comment]
> the entire point of this is that we don't call reboot until after the ssh session closes, which means that we won't catch errors.

Right, good point! And we'll have the output in /var/log anyway. Let's give this a try

File: machine/machine_core/ssh_connection.py [hunk 1]
```diff
@@ -146,6 +146,12 @@ def wait_reboot(self, timeout_sec=180):
         else:
             raise exceptions.Failure("Timeout waiting for system to reboot properly")
 
+    def reboot(self, timeout_sec=180):
+        # run cat to let spawned process wait for stdin to close (when m.execute() finishes)
```
Comment (allisonkarlitskaya) [inline machine/machine_core/ssh_connection.py:150]: This comment is no longer accurate.

File: machine/machine_core/ssh_connection.py [hunk 2]
```diff
@@ -253,7 +259,7 @@ def __execution_opts(self, direct=False):
 
     def execute(self, command=None, script=None, input=None, environment={},
                 stdout=None, quiet=False, direct=False, timeout=120,
-                ssh_env=["env", "-u", "LANGUAGE", "LC_ALL=C"]):
+                ssh_env=["env", "-u", "LANGUAGE", "LC_ALL=C"], ignore_exit_status=False):
```
Comment (allisonkarlitskaya) [inline machine/machine_core/ssh_connection.py:262]: There is a very strong tradition for this parameter to be called `check`.

**Review by allisonkarlitskaya — CHANGES_REQUESTED** [review]
(blocking because of the wrong comment.  the naming thing is a bikeshed.)

File: machine/machine_core/ssh_connection.py (hunk 1 above)
Comment (martinpitt) [reply machine/machine_core/ssh_connection.py]: Dropped.

**Comment by martinpitt** [comment]
Tests in https://github.com/cockpit-project/cockpit/pull/17089 are happy, so I think this is what I want to go with.

**Pull request merged**